src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Threading;
+using Microsoft.VisualStudio.Composition;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal sealed class ExportProviderBuilder
+{
+    private const string CatalogSuffix = ".mef-composition";
+
+    // For testing purposes, track the last cache write task.
+    private static Task? s_cacheWriteTask;
```

Comment:
Can we put this bit back in the language server layer, if it's only being used there?